pkg/dashboard/controller.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// It gets all the available foreign clusters connected to the local cluster and calculate
+// incoming and outgoing resources.
+func getForeignClusters(ctx context.Context, cl client.Client) (*[]ClusterDto, error) {
+	foreignClusterList := &discoveryv1alpha1.ForeignClusterList{}
+	err := cl.List(ctx, foreignClusterList)
+	if err != nil {
+		klog.Warningf("error retrieving foreign clusters: %s", err)
+		return nil, err
+	}
+
+	podMetricsList := &metricsv1beta1.PodMetricsList{}
+	err = cl.List(ctx, podMetricsList, client.MatchingLabels{
+		liqoconsts.LocalPodLabelKey: "true",
+	})
+	if err != nil {
+		klog.Warningf("error retrieving pod metrics: %s", err)
+		return nil, err
+	}
+	podMetricsMap := podMetricListToMap(podMetricsList.Items)
+
+	var clusters []ClusterDto
+	for i := range foreignClusterList.Items {
+		clusterDto := fromForeignCluster(&foreignClusterList.Items[i])
+
+		if isPeeringEstablished(clusterDto.OutgoingPeering) {
+			klog.V(5).Infof("Calculating outgoing resources for cluster %s", clusterDto.clusterID)
+			outgoingResources, err := calculateOutgoingResources(ctx, cl, clusterDto.clusterID, podMetricsMap)
+			if err == nil {
+				clusterDto.OutgoingResources = outgoingResources
+			} // otherwise, the outgoing resources are not calculated so they are nil
+		}
+
+		if isPeeringEstablished(clusterDto.IncomingPeering) {
+			incomingResources, err := calculateIncomingResources(ctx, cl, clusterDto.clusterID)
+			if err == nil {
+				clusterDto.IncomingResources = incomingResources
+			} // otherwise, the incoming resources are not calculated so they are nil
+		}
+
+		// in this moment clusters without any resource are also added to the list but we can decide to filter them
+		clusters = append(clusters, *clusterDto)
+	}
+
+	return &clusters, nil
+}
+
+// CalculateOutgoingResources The outgoing˙ resources aren't calculated, but they are simply retrieved from the metrics of the virtual node. The
+// clusterID identifies the virtual node by the label liqo.io/remote-cluster-id=clusterID.
+func calculateOutgoingResources(ctx context.Context, cl client.Client, clusterID string,
+	shadowPodsMetrics map[string]*metricsv1beta1.PodMetrics) (*ResourceMetrics, error) {
+	resourceOfferList := &sharingv1alpha1.ResourceOfferList{}
+	if err := cl.List(ctx, resourceOfferList, client.MatchingLabels{
+		liqoconsts.ReplicationOriginLabel: clusterID,
+	}); err != nil {
+		klog.V(5).Infof("error retrieving resourceOffers: %s", err)
+		return nil, err
+	}
+	if len(resourceOfferList.Items) != 1 {
+		return nil, fmt.Errorf("expected exactly one element in the list of ResourceOffer but got %d", len(resourceOfferList.Items))
+	}
+
+	podList, err := getOutgoingPods(ctx, cl, clusterID)
+	if err != nil {
+		klog.V(5).Infof("error retrieving outgoing pods: %s", err)
+		return nil, err
+	}
+
+	currentPodMetrics := []metricsv1beta1.PodMetrics{}
+	for i := range podList {
+		singlePodMetrics := shadowPodsMetrics[podList[i].Name]
+		if singlePodMetrics != nil {
```

Comment:
```suggestion
		singlePodMetrics, found := shadowPodsMetrics[podList[i].Name]
		if found {
```

Nit. Slightly clearer to me.